languages/csharp/AppEncryption/AppEncryption/SessionFactory.cs

```diff
@@ -128,18 +160,172 @@ internal Partition GetPartition(string partitionId)
             return new Partition(partitionId, serviceId, productId);
         }
 
-        private EnvelopeEncryptionJsonImpl GetEnvelopeEncryptionJson(string partitionId)
+        /// <summary>
+        /// Atomically acquires a shared <code>CachedSession</code> from the session cache for the
+        /// <code>partitionId</code>, creating a new one using the given function if needed. This is used to track the
+        /// number of concurrent users so cache eviction policies don't remove an entry while it's still potentially in
+        /// use.
+        /// </summary>
+        ///
+        /// <returns>The cached session that's mapped for the given <code>partitionId</code></returns>
+        ///
+        /// <param name="createSessionFunc">the function to create a new session if there is no current mapping</param>
+        /// <param name="partitionId">the partition id for a session</param>
+        private CachedSession AcquireShared(
+            Func<EnvelopeEncryptionJsonImpl> createSessionFunc, string partitionId)
+        {
+            SemaphoreSlim getCachedItemLock = semaphoreLocks.GetOrAdd(partitionId, k => new SemaphoreSlim(1, 1));
```

Comment:
Seems to me that a semaphore with an initial and maximum requests set to 1 would be functionally equivalent to `lock(getCachedItemLock)`. Is there a benefit to using a `SemaphoreSlim(1, 1)` here that I might be missing? If not, replacing it with a simple `Object()` with `lock` will simplify the code a bit and likely improve performance.

Reply:
i remembered that semaphoreslim provided better performance than simple lock, but i think in case of setting the limit to `1`, using just a `lock` would be a better option. Would make these changes soon.